### PR TITLE
Episode05 Preprocessors

### PR DESCRIPTION
**Pull request by rswamina**
First Pull request for Chapter 5 (Preprocessors). Work is ongoing on this chapter.
* * *

- [x] Read [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md).
- [x] Create an [issue](https://github.com/ESMValGroup/tutorial/issues) to discuss your idea. This allows your contributions to be incorporated into the tutorial.

## Tasks

- [x] Give this pull request a descriptive title.
- [x] If you are contributing to existing lesson materials, please make sure the content conforms to the `Lesson development` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [x] If you are making a new episode, please make sure the content conforms to the `Lesson organization` and `Lesson formatting` sections in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] Preferably Codacy checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why.
- [x] Preview changes on your machine before pushing them to GitHub by running `make serve`, alternatively `make docker-serve`. Please see the `Previewing your changes locally` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) for installation instructions.
- [ ] All code instructions have been tested.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes {https://github.com/ESMValGroup/tutorial/issues/30} when successful.

### CODE REVIEW

File: _episodes/05-preprocessor.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
+---
+title: "Working with preprocessors"
+teaching: 20?
+exercises: 15?
+questions:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:5]: Maybe also add the question: "How can I combine different preprocessor functions?"

File: _episodes/05-preprocessor.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+- "How do I set up a preprocessor?"
+- "Can I use different preprocessors for different variables?"
+- "Can I use different datasets for different variables?"
+objectives:
+- "Create a recipe with multiple preprocessors"
+- "Use different preprocessors for different variables"
+- "Run a recipe with variables from different datasets"
+keypoints:
+- "A recipe can run different preprocessors at the same time."
```
Comment (hb326) [inline _episodes/05-preprocessor.md:14]: Is it actually "running" different preprocessors at the same time? Maybe rephrasing to "A recipe can work with different preprocessors at the same time"?

File: _episodes/05-preprocessor.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+> preprocessors:
+>   prep_timeseries:
+>     annual_statistics:
+>       operator: mean
+>~~~
+{: .source}
+
+For instance, the 'annual_statistics' with the  'operation: mean' argument preprocessor receives an iris cube, takes the annual average for each year of data in the cube, and returns the processed cube. 
+
+You could use several preprocessor steps listed in the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html). The standardised interface between the preprocessors allows them to be used modularly - like lego blocks. Almost any conceivable preprocessing order of operations can be performed using ESMValTool preprocessors.
```
Comment (hb326) [inline _episodes/05-preprocessor.md:41]: Maybe rephrase: "You can use one ore multiple preprocessor steps, listed in the [documentation], per precropcessor."

Comment (hb326) [inline _episodes/05-preprocessor.md:41]: What do you mean with "any conceivable preprocessing order of operations"? I thought there is an exact order how the different tasks in the preprocessor are handled?

Comment (hb326) [reply _episodes/05-preprocessor.md:41]: Never mind. I found the the "custom_order" command just now in your tutorial.  :smile:

File: _episodes/05-preprocessor.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+>
+>If you do not want your preprocessors to be applied in the [default order](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html), then add the following line to your preprocessor chain:
+>
+>~~~
+>    custom_order: true
+>~~~
+>
+>The default preprocessor order is listed in the [ESMValCore preprocessor API page](>https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html).
+>
+> Also note that not preprocessor operations aren't always commutative - meaning that the order of operations matters. For instance, if you run the preprocessor 'extract_volume' to extract the top 100m of the ocean surface, then 'volume_statistics' to calculate the volume-weighted mean of the data, then your result will differ depending on the order of these two preprocessors. In fact, the 'extract_volume' preprocessor will fail if you try to run it on a 2D dataset.
```
Comment (hb326) [inline _episodes/05-preprocessor.md:53]: There is something off in the first half of the sentence. Either rephrase to "Also note that not all preprocessor operations are commutative..." or "Also note that preprocessor operations arent't always commutative..."

File: _episodes/05-preprocessor.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+> preprocessors:
+>   prep_timeseries_1:
+>     annual_statistics:
+>       Operator: mean
+>   prep_timeseries_2:
+>     annual_statistics:
+>       operator: mean
+>     area_statistics:
+>       operator: mean
+>     Depth_integration:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:94]: Should "Depth_integration" be capitalized here?

File: _episodes/05-preprocessor.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+## Using different preprocessors for different variables
+
+You could define different preprocessors with several preprocessor sections (setting different preprocessor names). In the variable section you call the specific preprocessor which should be applied.
+
+> ## Example
+>~~~YAML
+> preprocessors:
+>   prep_timeseries_1:
+>     annual_statistics:
+>       Operator: mean
```
Comment (hb326) [inline _episodes/05-preprocessor.md:88]: Should "Operator" be capitalized here?

File: _episodes/05-preprocessor.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+>   # Time series diagnostics
+>   # --------------------------------------------------
+>   diag_timeseries_temperature_1:
+>     description: simple_time_series>
+>     variables:
+>         timeseries_variable:
+>         short_name: thetaoga
+>         preprocessor: prep_timeseries_1
+>     scripts:
+>          timeseries_diag:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:107]: Intendation seems off...

File: _episodes/05-preprocessor.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+>         script: ocean/diagnostic_timeseries.py
+> 
+>   diag_timeseries_temperature_2:
+>     description: simple_time_series
+>     variables:
+>         timeseries_variable:
+>         short_name: thetao
+>         preprocessor: prep_timeseries_2
+>     scripts:
+>          timeseries_diag:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:117]: Intendation of "timeseries_diag" seems off...

File: _episodes/05-preprocessor.md [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+>         script: ocean/diagnostic_timeseries.py
+>~~~
+>{: .source}
+{: .solution}
+
+>## Challenge : How to write a recipe with multiple preprocessors
+> We now know that a recipe can have more than one diagnostic, variable or preprocessor. As we saw in the examples so far, we can group preprocessors with a single user defined name and can have more than one such preprocessor group in the recipe as well. Write two different preprocessors - one to regrid the data to a 1x1 resolution and the second preprocessor to mask out sea and ice grid cells before regridding to the saem resolution. In the second case, ensure you perform the masking first before regridding (hint: custom order your operations). Now, use the two preprocessors in different diagnostics within the same recipe. You may use any variable(s) of your choice. Once you succeed, try to add new datasets to the same recipe. Placeholders for the different components are provided below:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:124]: Typo in the word "saem".

File: _episodes/05-preprocessor.md [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+>
+>> ## Recipe
+>>
+>>~~~YAML
+>>
+>> datasets:
+>>   - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical,
```
Comment (hb326) [inline _episodes/05-preprocessor.md:131]: I think we should make sure that all episodes refer to the same project. Here you refer to CMIP6 data, but I think in episode 3 we refer to CMIP5 right now...

File: _episodes/05-preprocessor.md [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+>>      ensemble: r1i1p1f2} #single dataset as an example
+>> 
+>> preprocessors:
+>>   prep_map:
+>>     regrid:    #apply the preprocessor to regrid
+>>       target_grid: 1x1 # target resolution
+>>       scheme: linear  #how to interpolate for regridding
+>> 
+>>   prep_map_land:
+>>     custom_order: true #ensure order follows a given
```
Comment (hb326) [inline _episodes/05-preprocessor.md:187]: It seems like there are words missing in the comment line after "given"

File: _episodes/05-preprocessor.md [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+>>   diag_simple_plot:
+>>     description: # preprocess a variable for a simple 2D plot
+>>     variables:
+>>       tas:  # surface temperature
+>>         preprocessor: prep_map
+>>         mip: Amon
+>>         grid: gn #can change for variables from the same model
+>>         start_year: 1970 
+>>         end_year: 2000
+>>         additional_datasets:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:208]: "additional_datasets" were not mentioned in the script above. Maybe take it out here as well?

File: _episodes/05-preprocessor.md [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+>>   diag_land_only_plot:
+>>     description: #preprocess a variable for a 2D land only plot
+>>     variables:
+>>       tas:  # surface temperature
+>>       preprocessor: prep_map_land
+>>       mip: Amon
+>>       grid: gn #can change for variables from the same model
+>>       start_year: 1970 
+>>       end_year: 2000
+>>       additional_datasets:
```
Comment (hb326) [inline _episodes/05-preprocessor.md:220]: same as above: "additional_datasets" was not mentioned in the example script above, so maybe take it out here?

File: _episodes/05-preprocessor.md [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+>              version: v2.2, tier: 1} #dataset specific to this variable
+> 
+>       tas: #second variable is surface temperature
+>         preprocessor: prep_regrid
+>         mip: Amon
+>         grid: gn #can change for variables from the same model
+>         start_year: 1970  #some 30 year period
+>         end_year: 2000
+>         additional_datasets:
+>           - {dataset: HadCRUT4, project: OBS, type: ground,
```
Comment (hb326) [inline _episodes/05-preprocessor.md:271]: Do we want to mention that the keys that are necessary for identifying observational data is different, depending on their project? For example, for "obs4mips" data you have to specify "level" and "version", whereas for "OBS" you need "type" and "version", but no information about "level"...

File: _episodes/05-preprocessor.md [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+> 
+>       tas: #second variable is surface temperature
+>         preprocessor: prep_regrid
+>         mip: Amon
+>         grid: gn #can change for variables from the same model
+>         start_year: 1970  #some 30 year period
+>         end_year: 2000
+>         additional_datasets:
+>           - {dataset: HadCRUT4, project: OBS, type: ground, 
+>              version: 1, tier: 2} #dataset specific to the temperature var
```
Comment (hb326) [inline _episodes/05-preprocessor.md:272]: I think the last word in the comment section should be "variable"?

File: _episodes/05-preprocessor.md (hunk 14 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:271]: Not sure if that is being covered in Chapter 4. WIll check.

File: _episodes/05-preprocessor.md [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+> 
+>     scripts: null
+>~~~
+>{: .source}
+{: .solution}
+
+> ## How to find what CMIP data is available?
+>
+> [CMIP5](https://pcmdi.llnl.gov/mips/cmip5/index.html) and [CMIP6](https://pcmdi.llnl.gov/CMIP6/Guide/dataUsers.html) data obey the [CF-conventions](http://cfconventions.org/). Available variables could be found under the [CMIP5 data request](https://pcmdi.llnl.gov/mips/cmip5/docs/standard_output.pdf?id=28) and the [CMIP6 Data Request](http://clipc-services.ceda.ac.uk/dreq/index.html).
```
Comment (hb326) [inline _episodes/05-preprocessor.md:281]: Should be rephrased to "Available variables can be found under the ..."? "Could" sounds here so speculative...

File: _episodes/05-preprocessor.md (hunk 1 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:5]: Done.

File: _episodes/05-preprocessor.md (hunk 2 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:14]: When recipes run in paralllel, do they only run the diagnostics in parallel or do preprocessor functions also go in parallel. At any rate, the change suggested by @hb326 is valid so will make it but good to clarify the other point on whether different preprocessors run at the same time.

File: _episodes/05-preprocessor.md (hunk 3 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:41]: At this point we haven't talked about grouping preprocessor functions or using multiple preprocessors so have changed it accordingly to - "You may use one or more of several preprocessors listed in the documentation.."

Comment (rswamina) [reply _episodes/05-preprocessor.md:41]: There is an order in which preprocessors are handled but "any conceivable" here implies "any order you can think of" and that is true if you use custom_order. I found two locations in the Readme docs where there is information for the default order and it is a bit odd that they list the preprocessors quite differently. The first is what showed up with a Google search. This could have implications - for example, the need to regrid before multimodel means is there if we assume the first list of ordering whereas it doesn't need to be for the second list. We have provided the link to the second page but I believe we have to regrid before doing multimodel means, so which is correct? And should we have two locations in the readme docs for this showing different orders? I am calling the attendtion of @hb326 and @LisaBock ('s) to this.
https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html
https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html

File: _episodes/05-preprocessor.md (hunk 4 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:53]: Fixed that and the next sentence as well.

File: _episodes/05-preprocessor.md (hunk 6 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md]: No. Fixed.

File: _episodes/05-preprocessor.md (hunk 7 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:107]: Fixed.

File: _episodes/05-preprocessor.md (hunk 3 above)
Comment (LisaBock) [reply _episodes/05-preprocessor.md:41]: I think the second one is the correct order we should link to. In the first link it says that it is "roughly following the default order". so maybe not totally correct as it is sorted in groups.
And as I read, also in the second list `multi_model_statistics` is listed after `regrid`. It is not clear to me what you mean...

File: _episodes/05-preprocessor.md (hunk 13 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md:220]: Removed.

File: _episodes/05-preprocessor.md (hunk 3 above)
Comment (rswamina) [reply _episodes/05-preprocessor.md]: Ok..I see. I think I mistook regrid_time when reading...so this is ok then?

**Comment by rswamina** [comment]
We haven't  listed any common errors. What should we do about that?

**Comment by LisaBock** [comment]
> 
> 
> We haven't listed any common errors. What should we do about that?

@rswamina I think we should skip it for the moment as we want to bring all together today...

**Comment by rswamina** [comment]
> > We haven't listed any common errors. What should we do about that?
> 
> @rswamina I think we should skip it for the moment as we want to bring all together today...

Any more reviewers need or is this is for now?

**Comment by Peter9192** [comment]
> Any more reviewers needed or is this it for now? This is fine according to the guidance provided.

I'll quickly check technical stuff and whether all the conversations are resolved, and then merge if all is ok.

**Comment by rswamina** [comment]
> > Any more reviewers needed or is this it for now? This is fine according to the guidance provided.
> 
> I'll quickly check technical stuff and whether all the conversations are resolved, and then merge if all is ok.

Thanks, @Peter9192

File: _episodes/05-preprocessor.md (hunk 10 above)
Comment (Peter9192) [reply _episodes/05-preprocessor.md:132]: Let's try to align those sort of things in a next pass over the entire tutorial, once most of the content is merged.

**Comment by Peter9192** [comment]
Let me fix a few formatting issues with code blocks, then merge.

**Comment by Peter9192** [comment]
Codacy doesn't seem to trigger because Lisa is not a member of our organization.. Anyway, local check looks quite good. I'll merge! 
Thanks for the hard work on this!
